### PR TITLE
Working on Postgres Operator dependencies.

### PR DESCRIPTION
**Pull request by Jan-M**
Adding postgres operator with dependencies:

- [ ] new etcd cluster for postgres / patroni
- [ ] s3 backup for postgres
- [ ] iam role for pods
- [ ] iam role for operator

### CODE REVIEW

File: cluster/manifests/postgres-operator/configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: postgres-operator
+data:
+  service_account_name: operator
```
Comment (linki) [inline cluster/manifests/postgres-operator/configmap.yaml:6]: We don't pre-define the `operator` account anywhere. So it has to be created here as well. Alternatively we can also use `kube-system` namespace and the `system` user.

Comment (Jan-M) [reply cluster/manifests/postgres-operator/configmap.yaml:6]: Lets create it then and have an example with it. should not be the same as system imho

File: cluster/manifests/postgres-operator/configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  cluster_labels: application:spilo
+  cluster_name_label: version
+  pod_role_label: spilo-role
+  db_hosted_zone: {{ .ConfigItems.hosted_zone }}
```
Comment (linki) [inline cluster/manifests/postgres-operator/configmap.yaml:10]: TODO: add this to CLM

File: cluster/postgres-operator-dependencies.yaml [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      Policies:
+      - PolicyName: postgres-pod-policy
+        PolicyDocument:
+          Version: "2012-10-17"
+          Statement:
+            - Effect: "Allow"
+              Action:
+              - "s3:*"
+              Resource:
+              - "arn:aws:s3:::{{ Arguments.postgres_backup_bucket }}"
```
Comment (Jan-M) [inline cluster/postgres-operator-dependencies.yaml:161]: This also needs to come from CLM or we need to build a bucket name here

File: cluster/manifests/postgres-operator/credentials.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+apiVersion: "zalando.org/v1"
+kind: PlatformCredentialsSet
+metadata:
+  name: postgresql-operator
+spec:
+  application: postgresql-operator
+  tokens:
+    read-only:
+      privileges:
+    cluster-registry-rw:
```
Comment (linki) [inline cluster/manifests/postgres-operator/credentials.yaml:10]: I'm wondering why the operator, which is per cluster, needs access to the cluster-registry?

File: cluster/manifests/postgres-operator/operator.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: postgres-operator
+  namespace: acid
```
Comment (linki) [inline cluster/manifests/postgres-operator/operator.yaml:5]: This is the only Kubernetes resource that has a namespace. Is that correct?

File: cluster/manifests/postgres-operator/operator.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        name: postgres-operator
+        application: postgres-operator
+        version: {{ version }}
```
Comment (linki) [inline cluster/manifests/postgres-operator/operator.yaml:13]: That must be a specific version

File: cluster/manifests/postgres-operator/operator.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    spec:
+      containers:
+      - env:
+        - name: MY_POD_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        - name: CONFIG_MAP_NAME
+          value: "postgres-operator"
+        image: pierone.stups.zalan.do/acid/postgres-operator:ba6529bba6529b
```
Comment (linki) [inline cluster/manifests/postgres-operator/operator.yaml:23]: `ba6529bba6529b` should probably be `ba6529b`

File: cluster/manifests/postgres-operator/configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  ready_wait_timeout: 30s
+  replication_username: replication
+  resource_check_interval: 3s
+  resource_check_timeout: 10m
+  resync_period: 5m
+  resync_period_pod: 5m
+  super_username: postgres
+  teams_api_url: https://teams.auth.zalando.com/api/
+  wal_s3_bucket: {{ .ConfigItems.postgres_backup_s3_bucket }}
+  kube_iam_role: app-spilo
```
Comment (alexeyklyukin) [inline cluster/manifests/postgres-operator/configmap.yaml:32]: the name better reflect that it belongs to the operator (we use postgres-operator in the db cluster).

File: cluster/manifests/postgres-operator/operator.yaml (hunk 7 above)
Comment (linki) [reply cluster/manifests/postgres-operator/operator.yaml:23]: also `registry.opensource.zalan.do`

File: cluster/postgres-operator-dependencies.yaml [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+      - PolicyName: AllowEC2Modify
+        PolicyDocument:
+          Version: "2012-10-17",
+          Statement:
+            - Effect: Allow
+              Action:
+                - ec2:Describe*
+                - ec2:Modify*
+
+  PostgresPodRole:
```
Comment (linki) [inline cluster/postgres-operator-dependencies.yaml:140]: This probably needs to be configured via an annotation on the deployment as well.

File: cluster/manifests/postgres-operator/credentials.yaml (hunk 4 above)
Comment (alexeyklyukin) [reply cluster/manifests/postgres-operator/credentials.yaml:10]: We neither read nor write it. I have no idea why it is there in the first place :-)

File: cluster/postgres-operator-dependencies.yaml [hunk 10]
```diff
@@ -0,0 +1,175 @@
+SenzaComponents:
```
Comment (mikkeloscar) [inline cluster/postgres-operator-dependencies.yaml:1]: Why do we need this stack?

It seems like it's quite out of scope for our base Kubernetes infrastructure to also provision extra *non-cluster* related stacks.

We already have the problem of manually updating the existing etcd stack when taupage is outdated etc. Adding another etcd is not making it simpler.

**Comment by mikkeloscar** [comment]
Closing as the postgres operator is already being deployed via CDP.
